lib/tasks/data/remove_transient_data.rb

```diff
@@ -12,8 +13,35 @@ class Session < ActiveRecord::Base
   def call
     Rails.logger.info("#{self.class.name}: processing")
 
-    Spree::StateChange.where("created_at < ?", RETENTION_PERIOD).delete_all
-    Spree::LogEntry.where("created_at < ?", RETENTION_PERIOD).delete_all
-    Session.where("updated_at < ?", RETENTION_PERIOD).delete_all
+    Spree::StateChange.where("created_at < ?", MEDIUM_RETENTION).delete_all
+    Spree::LogEntry.where("created_at < ?", MEDIUM_RETENTION).delete_all
+    Session.where("updated_at < ?", SHORT_RETENTION).delete_all
+
+    clear_old_cart_data!
+    clear_line_item_option_values!
+  end
+
+  private
+
+  def clear_old_cart_data!
+    old_carts = Spree::Order.where("state = 'cart' AND updated_at < ?", SHORT_RETENTION)
+    old_cart_line_items = Spree::LineItem.where(order_id: old_carts)
+    old_cart_adjustments = Spree::Adjustment.where(order_id: old_carts)
+
+    old_cart_adjustments.delete_all
+    old_cart_line_items.delete_all
+    old_carts.delete_all
+  end
+
+  def clear_line_item_option_values!
+    ActiveRecord::Base.connection.execute <<-SQL
+      DELETE FROM spree_option_values_line_items
+      WHERE line_item_id IN (
+        SELECT line_item_id FROM spree_option_values_line_items
+        LEFT OUTER JOIN spree_line_items
+        ON spree_option_values_line_items.line_item_id = spree_line_items.id
+        WHERE spree_line_items.id IS NULL
+      );
+    SQL
```

Comment:
I quickly checked the query plan for this (assuming this DELETE will have a similar cost of a SELECT):

```sql
                                                                   QUERY PLAN                                                                   
------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=5696.25..5778.98 rows=235 width=0)
   ->  HashAggregate  (cost=5695.96..5698.31 rows=235 width=4)
         Group Key: spree_option_values_line_items_1.line_item_id
         ->  Hash Anti Join  (cost=3199.86..5695.37 rows=235 width=4)
               Hash Cond: (spree_option_values_line_items_1.line_item_id = spree_line_items.id)
               ->  Seq Scan on spree_option_values_line_items spree_option_values_line_items_1  (cost=0.00..1210.40 rows=82940 width=4)
               ->  Hash  (cost=1842.05..1842.05 rows=82705 width=4)
                     ->  Seq Scan on spree_line_items  (cost=0.00..1842.05 rows=82705 width=4)
   ->  Index Only Scan using index_option_values_line_items_on_line_item_id on spree_option_values_line_items  (cost=0.29..0.33 rows=1 width=4)
         Index Cond: (line_item_id = spree_option_values_line_items_1.line_item_id)
(10 rows)
```

and I'm afraid this put our DBs in a hard situation. That `Seq Scan on spree_line_items  (cost=0.00..1842.05 rows=82705 width=4)` bit needs some deeper investigation and perhaps some indexing. Have you checked writing replacing IN with joins?

Reply:
I've created a basic AR model to map the join table (we have a similar one for `Spree::ProductOptionType`) and I've moved things around a bit. `#delete_all` is now being called directly on a set of records in that join table, and the column in  question (`line_item_id`) already has an index on it :+1: